addons/web_editor/static/src/js/editor/snippets.editor.js

```diff
@@ -223,6 +223,17 @@ var SnippetEditor = Widget.extend({
         }
         return _t("Block");
     },
+    isUpToDate: function () {
+        const targetClass = [...this.$target[0].classList].filter(x => /\bs_./g.test(x))[0];
+        if (!targetClass) {
+            return true;
+        }
+        const referenceSection = $('#o_scroll')[0].querySelector(`.oe_snippet > .${targetClass}`);
+        if (!referenceSection) {
+            return true;
+        }
```

Comment:
When you don't find the reference of a snippet, you say it is up-to-date? Should not it be the opposite?

Reply:
If I do not find the snippet it is because the selected element it is not a top level snippet (versionable)  but rather an element which appear to have an s_ class and a selector with related snippet options.

Reply:
There is a flow in identifying which snippet is which. I think using the main s_ class is our best option now. 
Else we will need to add an identifier to all snippets and not change it from version to version. If we want a more robust system we should start to do that now and use the class system as a fallback for V13 and older snippets.

Reply:
I spoke with @rdeodoo about this.

In fact, we should have 3 types of classes:
A. s_xxx: for snippets
B. s_xxx_yyy: for snippet components / specific options
C. o_zzz: for common option classes and others

This will allow to make the up-to-date function work and enforce that convention:
1. Get the first s_ class of the target element (like you did)
1. On editor start, compute a list with all up-to-date snippets className (the first s_ class)
2. When a snippet portion is enabled in the left panel, get its first s_ class (like you did)
    a. Check if there is an exact match in the classNames found at point 1, if so check the version numbers
    b. Check if the className *startsWith* one of the classNames found at point 1, if so consider it up-to-date
    c. Otherwise, consider it outdated -> this will detect removed snippet *and* snippet using the wrong convention and that should thus be fixed for clarity

The other solution would be to forget about convention B and convert all non-main snippet classes with `o_`... maybe it will be better after all...

@JKE-be @rdeodoo Thoughts ?

Reply:
Ok, discussed with JKE in fact there is no problem, whatever our convention:
- If the s_ class you find is used by a main snippet, no problem: compare the version numbers
- If the s_ class that you find is not used by a main snippet consider it up-to-date (as you did) **but not if there is no parent editor anymore**. So if one day we remove the form snippet, when you check if s_website_form_field is up-to-date, it will be true and it's ok since it is a component not a main snippet, but when you check if s_website_form is up-to-date, it should false because there is no snippet editor parent.

... EDIT: in fact no he does not work, if we remove an inner snippet, you would not consider used ones as deprecated as used inside an updated snippet :/

To think again tomorrow... @sdegueldre maybe some thoughts about this?

cc @JKE-be :/

Reply:
In fact maybe I have the solution: ask the main snippet class names (new and deprecated) to the server. The server can find them as it has assets views named assets_snippet_s_xxx_js/css_ddd. Problem is for snippets which did not have any js/css but I don't think there are many.... to think about tomorrow !

Reply:
It's possible to break compat with old snippets without touching the css or js (eg changing the structure of the xml with no css or js changes), I don't think using vcss and vjs as an "up-to-date metric" is a very robust solution.

We could use the same versioning system as for the css and publicWidgets, except we only keep the code for the latest version in this case. Options declare a version as a field at the top, and you bump the version number any time an incompatible change has been made. When a snippet is dropped for the first time, we create all its editors and tag it (and possible inner snippets) with the version number of each js option that has one. When clicking on a snippet and initializing its editor at a later point, we can check that the option's current version corresponds to the version on the element before creating the option (preventing any problem with snippets doing stuff in init/start/willStart).

Compared to the vjs and vcss approach, it also has the benefit that a lot of the changes in the css and js are not going to break option compatibility, and in those cases you wouldn't need to show a useless warning.

The main problem I can see with this approach is that we're going to have a *lot* of version numbers in the DOM. We can either use one `data-voption-[optionName]` attribute per option, or a single `data-voptions` where the value is a JSON object with the version numbers for each option but in either case it's gonna be a lot of cruft. This problem would go away entirely if the new editor allows us to manipulate this sort of thing inside a virtual DOM without having it render for users.

Reply:
Shouldn't the problem be taken in reverse?
When you deprecate a snippet, just write a migration patch that will add `s_outdated` on the snippet you are deprecating.
I don't think it should be too complicated to do and should be more robust.
That would also make the editor code easier as no need for logic to find outdated elements.

Reply:
@sdegueldre We are not going to version options, that's the whole point of our system: if a breaking change occurred, we increase vcss/vjs/vxml and we want to warn users about those *whether or not the options can still work on the old snippet* so that they know it is an outdated snippet (especially since they cannot report a bug about those as the bug fix answer will be: "there is a new snippet for that, use it").

@rdeodoo I guess we could do that, that's probably the best idea. That would mean writing a migration patch for the 80 snippets we deprecated though... but it has the advantage that we can deprecate anything we want in the DOM by just adding s_outdated on anything, not only snippet but base classes or common options. We'll talk about it with JKE tomorrow.

Reply:
> That would mean writing a migration patch for the 80 snippets we deprecated though

I am pretty sure it is a one time job, we probably just need to write some utils method in migration repo and then just do
```python
utils.deprecate_snippet('s_countdown')
utils.deprecate_snippet('s_banner')
...
```
that would find any view arch/html field/whatever which contains `s_countdown` and replace it by `s_countdown s_outdated` (just have to ensure we correctly target only `s_countdown` and not `s_countdown_child_class` but not hard, any of [`'`, `"`, ` `] just after `s_countdown` should do it. We don't even need to check if it was already outdated, worst case we add another `s_outdated` on the same line.
So no need of view parsing with etree/xml.

Reply:
Thought about that initially but I thought migration scripts only ran when you bumped the odoo version. After rereading it seems you just have to bump the module version :) All hail website version `1.69`!

> we increase vcss/vjs/vxml

So we *are* doing vxml now? Didn't get the memo, good to know.

Reply:
vxml for xml structure change of the snippet which would break options... but we will probably never have the case since we are most likely to change either vcss or vjs in those cases anyway.

Reply:
`vxml` is new? We never discussed it. That would work too if we implement it. I don't like versioning so I prefer the migration patch method if it works.

> Thought about that initially but I thought migration scripts only ran when you bumped the odoo version. 

Idk, the whole versioning was done for a rolling release purpose, no?

Reply:
I spoke of `vxml` with JKE while everyone could have heard but that was not important anyway we don't have the case :wink: 

In any case, let's not mix things up: rolling release or not, we need the versioning system to keep the old snippets working when we cannot migrate them. That was the whole point.

That being the point, vxml can be forgotten if we go with RDE's solution. But the inverse is not true @rdeodoo. If we mark the snippets whose compatibility is broken at migration, there is no problem. If we only rely on vxml,vjs,vcss then we can only detect the snippets that changed, not the snippets that are removed. So again, still think your solution is the best:

1. vjs and vcss so that old snippets keep working post migration, and only for that
2. deprecate snippet at migration to know about deprecated ones in the editor
3. no need of vxml since the old snippets which were only broken by an xml change are patched at migration by (2) (we just need to not forget about them)

Let's check with JKE tomorrow.